src/main/java/io/nats/client/impl/NatsMessage.java

```diff
@@ -40,74 +40,73 @@ static int copy(byte[] dest, int pos, String toCopy) {
         return pos;
     }
 
+    private static String PUB_SPACE = NatsConnection.OP_PUB + " ";
+    private static String SPACE = " ";
+
     // Create a message to publish
-    NatsMessage(String subject, String replyTo, byte[] data) {
+    NatsMessage(String subject, String replyTo, byte[] data, boolean utf8mode) {
         this.subject = subject;
         this.replyTo = replyTo;
         this.data = data;
         
-        /* If we ever want to support UTF-8 subjects ...
-        private static String PUB_SPACE = NatsConnection.OP_PUB + " ";
-        private static String SPACE = " ";
-        StringBuilder protocolStringBuilder = new StringBuilder((2 * subject.length()) + 20); // guess a size based on replyTO = subject
-        protocolStringBuilder.append(PUB_SPACE);
-
-        protocolStringBuilder.append(subject);
-        protocolStringBuilder.append(SPACE);
-
-        if (replyTo != null) {
-            protocolStringBuilder.append(replyTo);
+        if (utf8mode) {
+            StringBuilder protocolStringBuilder = new StringBuilder((2 * subject.length()) + 20); // guess a size based on replyTO = subject

```

Comment:
I guess it is just an estimation and size will grow as needed, but strange that you use 2*subject. You could have something like:
```
int subjectSize = subject.length() * 2;
int replySize = (replyTo != null) ? replyTo.length() * 2 : 0;
StringBuilder protocolStringBuilder = new StringBuilder(4 + subjectSize + 1 + replySize + 1);
```

Reply:
made the change.